src/dapp/walletProvider.ts

```diff
@@ -285,6 +287,70 @@ export class WalletProvider implements IDappProvider {
         });
     }
 
+    /**
+     * Packs a {@link Transaction} and fetches correct redirect URL from the wallet API. Then redirects
+     *   the client to the sign transaction hook
+     * @param transaction
+     * @param options
+     */
+    async signTransaction(transaction: Transaction, options?: {callbackUrl?: string}): Promise<Transaction> {
+        if (!this.mainFrame) {
+            throw new Error("Wallet provider is not initialised, call init() first");
+        }
+
+        const {contentWindow} = this.mainFrame;
+        if (!contentWindow) {
+            throw new Error("Wallet provider is not initialised, call init() first");
+        }
+
+        return new Promise<Transaction>((resolve, reject) => {
+            let plainTransaction = transaction.toPlainObject();
+
+            // We adjust the fields, in order to make them compatible with what the wallet expected
+            plainTransaction["nonce"] = transaction.getNonce().valueOf();
+            plainTransaction["data"] = transaction.getData().valueOf().toString();
+            plainTransaction["value"] = transaction.getValue().toString();
+            plainTransaction["gasPrice"] = transaction.getGasPrice().valueOf();
+            plainTransaction["gasLimit"] = transaction.getGasLimit().valueOf();
+            console.log("postMessage", DAPP_MESSAGE_SIGN_TRANSACTION_URL, plainTransaction);
+
+            contentWindow.postMessage({
+                type: DAPP_MESSAGE_SIGN_TRANSACTION_URL,
+                data: {
+                    transaction: plainTransaction
+                }
+            }, this.walletUrl);
+
+            const timeout = setTimeout(_ => reject('sign transaction url not responding'), 5000);
+            const signTransactionUrl = (ev: IDappMessageEvent) => {
```

Comment:
also here, can you extract into getTransactionUrl

Reply:
I don't get why - this is already a function. It's encapsulated in the class method because it does not have a point to be an actual method itself. It follows the same pattern in all functions :)

Reply:
it was for readability, it easier to follow the code when there are smaller methods